File: src/main/java/de/spries/fleetcommander/model/facade/PlayerSpecificGame.java
package de.spries.fleetcommander.model.facade;

import java.util.List;

import de.spries.fleetcommander.model.ai.ComputerPlayer;
import de.spries.fleetcommander.model.ai.behavior.AggressiveFleetStrategy;
import de.spries.fleetcommander.model.ai.behavior.DefaultBuildingStrategy;
import de.spries.fleetcommander.model.ai.behavior.DefaultProductionStrategy;
import de.spries.fleetcommander.model.ai.behavior.SearchEnemyHomePlanetFleetStrategy;
import de.spries.fleetcommander.model.core.Game;
import de.spries.fleetcommander.model.core.Game.Status;
import de.spries.fleetcommander.model.core.Player;
import de.spries.fleetcommander.model.core.TurnEvents;

public class PlayerSpecificGame {

	private Game originalGame;
	private Player viewingPlayer;

	public PlayerSpecificGame(Game originalGame, Player viewingPlayer) {
		this.originalGame = originalGame;
		this.viewingPlayer = viewingPlayer;
	}

	public int getId() {
		return originalGame.getId();
	}

	public void addComputerPlayer() {
		long numComputerPlayers = getOtherPlayers().stream().filter(p -> !p.isHumanPlayer()).count();
		String name = "Computer " + (numComputerPlayers + 1);
		ComputerPlayer player = new ComputerPlayer(name, new DefaultBuildingStrategy(), new SearchEnemyHomePlanetFleetStrategy(), new DefaultProductionStrategy());
		originalGame.addPlayer(player);
	}

	public void addHumanPlayer(String playerName) {
		originalGame.addPlayer(new Player(playerName));
	}

	public Status getStatus() {
		return originalGame.getStatus();
	}

	public int getTurnNumber() {
		return originalGame.getTurnNumber();
	}

	public PlayerSpecificTurnEvents getPreviousTurnEvents() {
		TurnEvents events = originalGame.getPreviousTurnEvents();
		if (events != null) {
			return new PlayerSpecificTurnEvents(events, viewingPlayer);
		}
		return null;
	}

	public void start() {
		originalGame.start(viewingPlayer);
	}

	public void endTurn() {
		originalGame.endTurn(viewingPlayer);
	}

	public void quit() {
		originalGame.quit(viewingPlayer);
	}

	public PlayerSpecificUniverse getUniverse() {
		if (originalGame.getUniverse() != null) {
			return PlayerSpecificUniverse.convert(originalGame.getUniverse(), viewingPlayer);
		}
		return null;
	}

	public OwnPlayer getMe() {
		return new OwnPlayer(viewingPlayer);
	}

	public List<OtherPlayer> getOtherPlayers() {
		List<Player> otherOriginalPlayers = Player.filterAllOtherPlayers(originalGame.getPlayers(), viewingPlayer);
		return OtherPlayer.convert(otherOriginalPlayers);
	}
}

File: src/main/kotlin/de/spries/fleetcommander/model/facade/PlayerSpecificGame.kt
package de.spries.fleetcommander.model.facade

import de.spries.fleetcommander.model.ai.ComputerPlayer
import de.spries.fleetcommander.model.ai.behavior.DefaultBuildingStrategy
import de.spries.fleetcommander.model.ai.behavior.DefaultProductionStrategy
import de.spries.fleetcommander.model.ai.behavior.SearchEnemyHomePlanetFleetStrategy
import de.spries.fleetcommander.model.core.Game
import de.spries.fleetcommander.model.core.Player

open class PlayerSpecificGame(private val originalGame: Game, private val viewingPlayer: Player) {

    fun getId() = originalGame.getId()

    fun getStatus() = originalGame.getStatus()

    fun getTurnNumber() = originalGame.getTurnNumber()

    fun getPreviousTurnEvents(): PlayerSpecificTurnEvents? {
        val events = originalGame.getPreviousTurnEvents()
        return if (events != null) {
            PlayerSpecificTurnEvents(events, viewingPlayer)
        } else null
    }

    fun getUniverse() = if (originalGame.getUniverse() != null) {
        PlayerSpecificUniverse.convert(originalGame.getUniverse()!!, viewingPlayer)
    } else null

    fun getMe() = OwnPlayer(viewingPlayer)

    fun getOtherPlayers() = Player.filterAllOtherPlayers(originalGame.getPlayers(), viewingPlayer)
            .map { OtherPlayer(it) }

    fun addComputerPlayer() {
        val numComputerPlayers = getOtherPlayers().count { p -> !p.isHumanPlayer() }
        val name = "Computer " + (numComputerPlayers + 1)
        val player = ComputerPlayer(name, DefaultBuildingStrategy(), SearchEnemyHomePlanetFleetStrategy(), DefaultProductionStrategy())
        originalGame.addPlayer(player)
    }

    fun addHumanPlayer(playerName: String) {
        originalGame.addPlayer(Player(playerName))
    }

    fun start() {
        originalGame.start(viewingPlayer)
    }

    fun endTurn() {
        originalGame.endTurn(viewingPlayer)
    }

    fun quit() {
        originalGame.quit(viewingPlayer)
    }
}
